tokens/usecase/tokens_usecase_test.go

```diff
@@ -539,5 +542,418 @@ func (s *TokensUseCaseTestSuite) TestGetMinPoolLiquidityCap() {
+			},
+		},
+		{
+			name: "Multiple valid denoms",
+			chainDenoms: map[any]any{
+				"denom1": struct{}{},
+				"denom2": struct{}{},
+			},
+			poolMetadata: domain.PoolDenomMetaDataMap{
+				"denom1": domain.PoolDenomMetaData{Price: osmomath.NewBigDec(10)},
+				"denom2": domain.PoolDenomMetaData{Price: osmomath.NewBigDec(6)},
+			},
+			expectedDenoms: []string{"denom1", "denom2"},
+			expectedMetadata: domain.PoolDenomMetaDataMap{
+				"denom1": domain.PoolDenomMetaData{Price: osmomath.NewBigDec(10)},
+				"denom2": domain.PoolDenomMetaData{Price: osmomath.NewBigDec(6)},
+			},
+		},
+		{
+			name: "Not valid denom",
+			chainDenoms: map[any]any{
+				"validdenom": struct{}{},
+			},
+			poolMetadata: domain.PoolDenomMetaDataMap{
+				"notvaliddenom": domain.PoolDenomMetaData{Price: osmomath.NewBigDec(10)},
+				"validdenom":    domain.PoolDenomMetaData{Price: osmomath.NewBigDec(6)},
+			},
+			expectedDenoms: []string{"validdenom"},
+			expectedMetadata: domain.PoolDenomMetaDataMap{
+				"validdenom": domain.PoolDenomMetaData{Price: osmomath.NewBigDec(6)},
+			},
+		},
+		{
+			name: "Not valid type denom",
+			chainDenoms: map[any]any{
+				1:        struct{}{},
+				"denom2": struct{}{},
+			},
+			poolMetadata: domain.PoolDenomMetaDataMap{
+				"denom1": domain.PoolDenomMetaData{Price: osmomath.NewBigDec(10)},
+				"denom2": domain.PoolDenomMetaData{Price: osmomath.NewBigDec(6)},
+			},
+			expectedDenoms: []string{"denom2"},
+			expectedMetadata: domain.PoolDenomMetaDataMap{
+				"denom2": domain.PoolDenomMetaData{Price: osmomath.NewBigDec(6)},
+			},
+		},
+		{
+			name:             "No valid denoms",
+			chainDenoms:      map[any]any{},
+			poolMetadata:     domain.PoolDenomMetaDataMap{},
+			expectedDenoms:   []string{},
+			expectedMetadata: domain.PoolDenomMetaDataMap{},
+		},
+	}
+
+	for _, tt := range testcases {
+		s.Run(tt.name, func() {
+			usecase := tokensusecase.NewTokensUsecase(nil, 0, nil)
+			usecase.UpdatePoolDenomMetadata(tt.poolMetadata)
+			for k, v := range tt.chainDenoms {
+				usecase.SetChainDenoms(k, v)
+			}
+
+			result := usecase.GetFullPoolDenomMetadata()
+			s.Require().Equal(tt.expectedMetadata, result)
+		})
+	}
+}
+
+// Tests the GetChainDenom function.
+func (s *TokensUseCaseTestSuite) TestGetChainDenom() {
+	testcases := []struct {
+		name           string
+		humanDenom     string
+		denomMap       map[string]any
+		expectedResult string
+		expectedError  error
+	}{
+		{
+			name:           "Valid human denom",
+			humanDenom:     "validDenom",
+			denomMap:       map[string]any{"validdenom": "chainDenom"},
+			expectedResult: "chainDenom",
+			expectedError:  nil,
+		},
+		{
+			name:           "Invalid human denom - not found",
+			humanDenom:     "invalidDenom",
+			denomMap:       map[string]any{"validdenom": "chainDenom"},
+			expectedResult: "",
+			expectedError:  fmt.Errorf("chain denom for human denom (invaliddenom) is not found"),
```

Comment:
For this and other tests, let's please move the errors to `domain/errors.go` so that if we change an error message, we only need to do it in one place.

Additionally, sometimes we might require reusing the same error when testing some of the upstream methods.

See an example [here](https://github.com/osmosis-labs/sqs/blob/48713b07f622e2c377d37598ab77f4b6d7e18ede/router/usecase/optimized_routes_test.go#L345)

Reply:
I have created distinct error types as per [your example](https://github.com/osmosis-labs/sqs/blob/48713b07f622e2c377d37598ab77f4b6d7e18ede/router/usecase/optimized_routes_test.go#L345), however instead of moving to `domain/errors.go` I left them in  `tokens/usecase/errors.go`,  since they seems too verbose for domain at this point, please let me know if I should move to `domain/errors.go`. In general it seems that number of errors started growing and my concern is that we can end up exposing too much implementation details and making API less flexible, IMHO it would be useful to came up with more general error messages, at this point I wanted to avoid refactoring actual messages in case to clients relies on them.

Reply:
Discussed with Roman, we should also add some more format documentation in CONTRIBUTING.md. ( Create a task ).

Reply:
[DATA-270](https://linear.app/osmosis/issue/DATA-270/refactor-and-or-formalise-errors-in-contributingmd-document) was created.